source/main/cardano/CardanoSelfnodeLauncher.js

```diff
@@ -0,0 +1,181 @@
+};
+
+let mockTokenMetadataServer = null;
+const platform = String(environment.platform);
+
+const CARDANO_WALLET_PORT = 8088;
+const CARDANO_WALLET_START_TIMEOUT = 60 * 1000; // 60 seconds | unit: milliseconds
+const CARDANO_WALLET_START_CHECK_INTERVAL = 500; // 500 ms | unit: milliseconds
+const SHELLEY_TEST_DATA = '../../utils/cardano/selfnode';
+const TOKEN_METADATA_REGISTRY = './utils/cardano/selfnode/token-metadata.json';
+const TOKEN_METADATA_SERVER_PORT = 65432;
+const TOKEN_METADATA_SERVER = `http://localhost:${TOKEN_METADATA_SERVER_PORT}/`;
+const TOKEN_METADATA_SERVER_PROCESS_NAME =
+  platform === 'win32'
+    ? 'mock-token-metadata-server.exe'
+    : 'mock-token-metadata-server';
+
+export async function CardanoSelfnodeLauncher(
+  selfnodeOptions: SelfnodeOptions
+): Promise<{
+  node: Selfnode,
+  replyPort: number,
+}> {
+  return new Promise(async (resolve, reject) => {
+    const {
+      selfnodeBin,
+      mockTokenMetadataServerBin,
+      processName,
+      onStop,
+    } = selfnodeOptions;
+    setupMockTokenMetadataServer(mockTokenMetadataServerBin);
+    const processList: Array<Process> = await find('port', CARDANO_WALLET_PORT);
+    const isSelfnodeRunning =
+      processList.length && processList[0].name === processName;
+    if (isSelfnodeRunning) {
+      logger.info('Cardano-node is already running...', {
+        selfnodeBin,
+        CARDANO_WALLET_PORT,
+        SHELLEY_TEST_DATA,
+        TOKEN_METADATA_SERVER,
+      });
+      const nodeProcess: Process = processList[0];
+      const node: Selfnode = setupSelfnode({
+        processName,
+        nodeProcess,
+        onStop,
+        connected: true,
+      });
+      resolve({ node, replyPort: CARDANO_WALLET_PORT });
+    } else {
+      logger.info('Starting cardano-node now...', {
+        selfnodeBin,
+        CARDANO_WALLET_PORT,
+        SHELLEY_TEST_DATA,
+        TOKEN_METADATA_SERVER,
+      });
+      const nodeProcess: ChildProcess = spawn(selfnodeBin, [], {
+        env: {
+          ...process.env,
+          CARDANO_WALLET_PORT,
+          SHELLEY_TEST_DATA,
+          TOKEN_METADATA_SERVER,
+        },
+        detached: true, // allows Daedalus to exit independently of selfnode (1/3)
+        stdio: 'ignore', // allows Daedalus to exit independently of selfnode (2/3)
+      });
+      nodeProcess.unref(); // allows Daedalus to exit independently of selfnode (3/3)
+      const node: Selfnode = setupSelfnode({
+        processName,
+        nodeProcess,
+        onStop,
+        connected: false, // selfnode is kept in disconnected state until it starts responding
+      });
+      tcpPortUsed
+        .waitUntilUsed(
+          CARDANO_WALLET_PORT,
+          CARDANO_WALLET_START_CHECK_INTERVAL,
+          CARDANO_WALLET_START_TIMEOUT
+        )
+        .then(
+          () => {
+            node.connected = true;
+            resolve({ node, replyPort: CARDANO_WALLET_PORT });
+          },
+          (exitStatus) => {
```

Comment:
Just minor, but for consistency, using `catch` would be preferred.

Reply:
Great catch @yakovkaravelov - fixed in https://github.com/input-output-hk/daedalus/pull/2462/commits/157914e9f6f1112d7e08bbf876e0b2fa473eb624